### PR TITLE
Issue #17483: sort feature flag

### PR DESCRIPTION
**Pull request by DavidAn830**
## Hey, I just made a Pull Request!

Sorted feature flags so that active flags show up on the top of the list. This new list is also organized in its registered order so that it creates less confusion as flags travel through the list.

https://github.com/backstage/backstage/issues/17483

When no flags are active
Screenshot 2023-04-26 at 12 55 29 PM

When some flags are active
Screenshot 2023-04-26 at 12 55 16 PM

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by DavidAn830** [comment]
it looks like everyone is having kubernetes-backend api-report bugs and E2E Linux errors...

**Comment by DavidAn830** [comment]
@Rugvip Hey! I synced the fork with the master and now it passes all checks. Please take a look when you have time and let me know any suggestions you might have!

File: packages/core-app-api/api-report.md [hunk 1]
```diff
@@ -445,6 +445,8 @@ export class LocalStorageFeatureFlags implements FeatureFlagsApi {
   // (undocumented)
   getRegisteredFlags(): FeatureFlag[];
   // (undocumented)
+  getSortedFlags(): FeatureFlag[];
```
Comment (Rugvip) [inline packages/core-app-api/api-report.md:448]: Let's not add this to the core API, and instead just sort them in `UserSettingsFeatureFlags`

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you! 👍 

I'd like to avoid changes to the core API, they're not really worth it in this case.

I tried it out in the preview deployment as well and I do think the switches jumping around on toggle is a bit too jarring. Let's make sure the order stays the same while on the page, and only update in on the initial load.

**Comment by DavidAn830** [comment]
@Rugvip Hey! I received your feedback on the re-render part and applied it in this PR! It should only sorts on the first time loading now. I also removed the api change as you requested and it fails the verify check. Is there a way to prevent this?

**Review by Rugvip — CHANGES_REQUESTED** [review]
You'll need to move the sorting logic into the `UserSettingsFeatureFlags` component itself, or adjacent helper. The changes in the other packages need to be dropped.

**Comment by DavidAn830** [comment]
@Rugvip , I was trying a few different ways to accomplish the same change without adding any changes to the core api. But it looks quite difficult to do it without changing core api. The problem is that everytime a flag is toggled the component rerenders, it rerenders with the unsorted `registeredFeatureFlags` variable. I think a setter for `registeredFeatureFlags` in `LocalStorageFeatureFlags` would help solve this issue and it makes sense to have one in a class. Let me know your thoughts.

**Comment by freben** [comment]
But your code can sort immediately before returning the JSX, right? As in, performing an array sort on every render, not to store it in a state.

**Comment by DavidAn830** [comment]
@freben I will give it another shot but I think the problem with that is that feature flags will jump around as they are toggled on and off, causing bad user experience as mentioned by @Rugvip. Also another concern I have with this is that the order of the flags in the list can be messed up easily.. And that, at least for me, hurt my eyes 😂

**Comment by freben** [comment]
Sure - you can still duplicate items to have a separate group of selected ones at the top too. Either way it can be part of the rendering code rather than the framework bits.

**Comment by DavidAn830** [comment]
Right! Thank you!

**Comment by DavidAn830** [comment]
removed the change in the core api and for a new approach, I used `useState` hook to prevent from featureflags order to be rendered every toggle action. Let me know any thoughts on it.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice, looks good! 👍 

Now all that's needed is a changeset and we can ship it, see https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets

**Comment by DavidAn830** [comment]
@Rugvip , Changeset is now added! Let me know anything else that needs to be worked on. Thanks for all the feedbacks and responses!

**Review by Rugvip — APPROVED** [review]
Awesome, thank you! :grin:

**Pull request merged**